### ISSUE TITLE
Filtering build triggers

### ISSUE THREAD

**Issue by johananl** [telia-oss/github-pr-resource#58]
Hi,

Thanks for contributing this useful resource to the Concourse ecosystem!

Recently we started experimenting with Concourse for some of our projects. We do lots of open-source work, which makes it necessary for us to somehow control the execution of CI jobs in PRs from untrusted sources. Namely, we would love to have a way to trigger builds for PRs only after they had been reviewed by a maintainer to prevent execution of malicious code on the CI platform.

Is there currently a way to filter build triggers for PRs based on, for example, the presence of a specific label or a comment from a maintainer?
If not - is this something that is going to happen in the future?

Thanks a lot!

**Comment by itsdalmo** [comment]
Hi, thanks for submitting the issue!

>Is there currently a way to filter build triggers for PRs based on, for example, the presence of a specific label or a comment from a maintainer?

Currently there is no such functionality in this resource, but it's definitely worth considering given your use-case! We'll need to investigate how to best achieve this however:

- labels: Malicious code can be introduced _after_ the label has been added.
- comments: Same as above, and we might need additional privileges in the `access_token` in order to determine whether the comment is from a maintainer or not.
- PR review: Same here, we'd need an approve for each commit to ensure that we don't run tests against malicious code.

A simple solution might be to add a `disable_forks` flag on the resource, and instead have two different pipelines, one for PRs made by people with write access to the repository, and one for PRs from forks, and then handle PRs from forks in a safer way:
- Use task definitions from `master` branch instead of the PR itself.
- No credentials handed to the task environment.
- Add `timeout:` for the task execution.

Would something like this work for your use case? Could you give some more details about what kind of security risk you are worried about when executing untrusted code?

**Comment by johananl** [comment]
Thanks for the prompt response @itsdalmo.

I think some filtering functionality existed in the [github-pullrequest-resource] resource. In any case, sure - this would definitely have to be considered carefully.

>labels: Malicious code can be introduced after the label has been added.

A possible implementation could be something like the following:

- Trigger: a PR is opened **or** new commits are pushed **or** a label is added.
- The resource looks for a specific label.
  - If the label doesn't exist, the pipeline doesn't run.
  - If the label exists it is removed and the pipeline runs.

>comments: Same as above, and we might need additional privileges in the access_token in order to determine whether the comment is from a maintainer or not.

I think this can be handled using a static list of maintainers in the resource's configuration, an "owners file" or anything similar. We can use the following logic for example:

- Trigger: - Trigger: a PR is opened **or** new commits are pushed **or** a comment is added.
- The resource looks for a specific comment from one of the maintainers which was added **after** the last push.
  - If such a comment doesn't exist, the pipeline doesn't run.
  - If such a comment exists, the pipeline runs.

Yes, `disable_forks` can also be useful in addition to the above.

In any case, I did not intend to propose a concrete implementation here as there are obviously lots of other things to think about, some of which might have to do with Concourse itself rather than this resource. There is an open [issue] about these concerns on the Concourse repo.

I might have more feedback (or more specific suggestions) after I had a chance to run our CI process with the resource. I leave it for you to decide if there is a point in leaving this issue open and/or discuss things further on this matter.

[github-pullrequest-resource]: https://github.com/jtarchie/github-pullrequest-resource#source-configuration
[issue]: https://github.com/concourse/concourse/issues/366

**Comment by jtarchie** [comment]
This functionality did exist in the original. It is also the reason I chose to retire the PR resource. Trying to maintain all the different ways someone uses PRs (via labels, comments, forks, brances, etc) became complex to maintain and difficult to minimize in API usage.

My _high_ recommendation is to have the use have a guard task that inspects the PR for their own restrictions. That way it is left to the user to do the correct API calls, spell incantations, etc.

**Comment by johananl** [comment]
Thanks for the input @jtarchie. So if I understand correctly, your recommendation is to let the pipeline always trigger, and check whatever we want *using pipeline code*, that is - without relying on the *resource* to do it for us?

**Comment by phillbaker** [comment]
As a related example - it might be helpful to filter out github PRs in the draft status. 

@jtarchie do you have any examples of the "guard task that inspects the PR for their own restrictions"?

**Comment by jtarchie** [comment]
I don't have examples.
The `hub` CLI is really powerful, as is the Github API documentation.
I stopped adding filtering options my previous resource because it was too much responsibility (and just hard) to account for everyone's _single use case_ of how they used PRs.

**Comment by iuriaranda** [comment]
I understand the problematic of implementing all those different filtering options in the resource, but I fail to see how such "guard task" should be set up. As far as I know there isn't a way to conditionally trigger steps in a job plan.

**Comment by jtarchie** [comment]
There is no conditional task.

What task are you running against your PR? Is it a single one or bunch of tasks?

If single, modify that one to inspect the PR for your conditions. If it matches run the task as usual. If doesn't just fail gracefully.

If multiple tasks, include a guard task on any invocation of PR. If it doesn't match fail as it is failure. It doesn't meet your requirements.

**Comment by iuriaranda** [comment]
ok, I see. It makes sense. Thank you for the clarification.

**Comment by itsdalmo** [comment]
> My high recommendation is to have the use have a guard task that inspects the PR for their own restrictions. That way it is left to the user to do the correct API calls, spell incantations, etc.

I tend to agree with this @jtarchie, but until resources v2 and the next generation of tasks come along I think a guard task would just increase complexity for both users and maintainers?
- Option A: `github-pr-resource` fetches labels as part of `get`, and a separate task filters the labels and throws an error if the label is not present:
   - Resource still has to have a new release everytime a new filter needs to be added.
   - All we really save is an `if`-statement in `in.go`.
- Option B: `github-pr-resource` does not return any additional information which can be used for filtering, it basically only passes the commit SHA/PR number and the task needs to get the filter information on its own:
  - The added API calls count toward the rate limit.
  - The task needs to re-implement things, like the logic for fetch a PR and handling credentials.
  - In general, tasks are harder to version and distribute.

And in the not-so-distant future I believe the revamps coming for resources and tasks (and spaces) means that `github-pr-resource` will undergo breaking changes as well. E.g., emitting `committed` as part of the version will no longer be necessary with spaces - and as part of those breaking changes, or when tasks get improved, we should be able to move the filters out into a separate `github-pr-filter-task` with minimal effort?

**Comment by itsdalmo** [comment]
Closing this since releasing #126 and now merging #104 (available on `dev` tag/soon to be released).
